### ISSUE TITLE
[TRT ENGINE] gpu preprocessing is slower than cpu

### ISSUE THREAD

**Issue by sh1man999** [ultralytics/ultralytics#23038]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Concurrency: 16, throughput: 23.4832 infer/sec, latency 681930 usec
Concurrency: 24, throughput: 27.828 infer/sec, latency 860303 usec
Concurrency: 32, throughput: 28.8983 infer/sec, latency 1101171 usec
Concurrency: 40, throughput: 29.4814 infer/sec, latency 1353208 usec

```

import json

import cv2
import numpy as np
import torch
from torch.utils.dlpack import from_dlpack, to_dlpack
import triton_python_backend_utils as pb_utils  # type: ignore


def letterbox(img: np.ndarray, new_shape: tuple[int, int]) -> tuple[np.ndarray, float, int, int]:
    shape = img.shape[:2]  # current shape [height, width]

    # Scale ratio (new / old)
    r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])

    # Compute padding
    new_unpad = round(shape[1] * r), round(shape[0] * r)
    dw, dh = (new_shape[1] - new_unpad[0]) / 2, (new_shape[0] - new_unpad[1]) / 2  # wh padding

    if shape[::-1] != new_unpad:  # resize
        img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_LINEAR)

    top, bottom = round(dh - 0.1), round(dh + 0.1)
    left, right = round(dw - 0.1), round(dw + 0.1)

    img = cv2.copyMakeBorder(img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=(114, 114, 114))
    return img, r, top, left

class TritonPythonModel:
    def initialize(self, args) -> None:
        model_config = json.loads(args["model_config"])
        params = model_config.get("parameters", {})
        imgsz_str = params.get("tgt_imgsz", {"string_value": "640,640"})["string_value"]
        self.tgt_h, self.tgt_w = map(int, imgsz_str.split(","))
        self.new_shape = (self.tgt_h, self.tgt_w)


    def execute(self, requests):
        responses = []

        for request in requests:
            original_image = pb_utils.get_input_tensor_by_name(request, "original_image").as_numpy()
            cropped_image, gain, pad_top, pad_left = letterbox(original_image, new_shape=self.new_shape)

            cropped_image = (cropped_image.astype(np.float32) / 255.0).transpose(2, 0, 1)
            cropped_image = np.expand_dims(cropped_image, axis=0)

            scale_pad_params = np.array([gain, pad_top, pad_left], dtype=np.float32)

            responses.append(
                pb_utils.InferenceResponse(
                    output_tensors=[
                        pb_utils.Tensor("original_image_scale_pad_params", scale_pad_params),
                        pb_utils.Tensor("vehicle_detection_preprocessed_images", cropped_image),
                    ]
                )
            )



        return responses

```

Concurrency: 16, throughput: 19.2327 infer/sec, latency 830547 usec
Concurrency: 24, throughput: 20.9264 infer/sec, latency 1144295 usec
Concurrency: 32, throughput: 21.2028 infer/sec, latency 1506107 usec
Concurrency: 40, throughput: 21.3971 infer/sec, latency 1855333 usec
```
import json

import cv2
import numpy as np
import torch
from torch.utils.dlpack import from_dlpack, to_dlpack



def letterbox_pytorch(img: torch.Tensor, new_shape: tuple[int, int]):
    """Выполняет ресайз и паддинг изображения с сохранением пропорций на GPU."""
    shape = img.shape[-2:]
    h, w = shape[0], shape[1]
    r = min(new_shape[0] / h, new_shape[1] / w)
    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
    dw, dh = (new_shape[1] - new_unpad[0]) / 2, (new_shape[0] - new_unpad[1]) / 2  # wh padding

    if shape[::-1] != new_unpad:
        img = F.interpolate(
            img.unsqueeze(0) if img.dim() == 3 else img,
            size=(new_unpad[1], new_unpad[0]),
            mode="bilinear",
            align_corners=False,
        )


class TritonPythonModel:

    def initialize(self, args) -> None:
        model_config = json.loads(args["model_config"])
        params = model_config.get("parameters", {})
        imgsz_str = params.get("tgt_imgsz", {"string_value": "640,640"})["string_value"]
        self.tgt_h, self.tgt_w = map(int, imgsz_str.split(","))
        self.new_shape = (self.tgt_h, self.tgt_w)

    def execute(self, requests):
        responses = []

        for request in requests:
            
            original_image = from_dlpack(pb_utils.get_input_tensor_by_name(request, "original_image").to_dlpack())
            img_chw = original_image.permute(2, 0, 1).float() / 255.0
            processed_img, gain, pad_top, pad_left = letterbox_pytorch(img_chw, new_shape=(self.tgt_h, self.tgt_w))

            scale_params = torch.tensor([gain, pad_top, pad_left], dtype=torch.float32, device="cuda")

            responses.append(
                pb_utils.InferenceResponse(
                    output_tensors=[
                        pb_utils.Tensor.from_dlpack("original_image_scale_pad_params", to_dlpack(scale_params)),
                        pb_utils.Tensor.from_dlpack("vehicle_detection_preprocessed_images", to_dlpack(processed_img)),
                    ]
                )
            )

        return responses

```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sh1man999, thank you for your interest in Ultralytics 🚀! This is an automated response to help us triage your issue, and an Ultralytics engineer will also assist soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In your case (TRT engine + GPU vs CPU preprocessing throughput/latency), an MRE should include the smallest runnable script/config that reproduces the slowdown, plus exact versions and the command(s) you used to measure concurrency/throughput.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ulalytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
This is expected for 640×640: Triton’s Python backend overhead (per-request loop, DLPack wrapping, kernel launch cost, and any implicit CPU↔GPU copies/syncs) can outweigh the benefit of doing `resize/pad/normalize` on GPU, and our published TensorRT benchmarks don’t include pre/post-processing time (only model inference) as noted in the docs for [TensorRT export](https://docs.ultralytics.com/integrations/tensorrt/) and the [Triton guide](https://docs.ultralytics.com/guides/triton-inference-server/). To confirm you’re actually preprocessing on GPU (and not on CPU via DLPack), can you share your `config.pbtxt` and the `torch.Tensor.device` after `from_dlpack(...)`; if the input arrives on CPU, `to("cuda")` will add a costly H2D copy and OpenCV CPU can easily win—so the usual fixes are keeping OpenCV preprocessing on CPU and feeding the TensorRT model directly, or moving preprocessing to a GPU-native Triton path (e.g., DALI/C++ backend) or batching preprocessing to amortize overhead.

**Comment by sh1man** [comment]
You're right, photo transfer goes through the processor.

```
name: "vehicle_detection_preprocess"
backend: "python"
max_batch_size: 0

instance_group [
  {
    count: 4
    kind: KIND_CPU
  }
]

parameters {
    key: "tgt_imgsz"
    value: {string_value: "640,640"}
}


input [
  {
    name: "original_image"
    data_type: TYPE_UINT8
    dims: [ -1, -1, 3 ]
  }
]

output [
  {
    name: "vehicle_detection_preprocessed_images"
    data_type: TYPE_FP32
    dims: [ 3, 640, 640]
  },
  {
    name: "original_image_scale_pad_params"
    data_type: TYPE_FP32
    dims: [3]
  }
]


```

**Comment by glenn-jocher** [comment]
Yep—that `config.pbtxt` forces the Python backend to run as `KIND_CPU`, so `original_image` arrives in host memory and any GPU “preprocess” will pay an H2D copy (plus sync/kernel-launch overhead), which is why CPU OpenCV wins at 640×640; if you truly need GPU preprocessing you’ll want an end-to-end GPU path (e.g., CUDA shared memory from a local client, or a GPU-native preprocessing backend like DALI/C++), and if you stick with H2D copies then batching (`max_batch_size>0` with fixed input shapes + dynamic batching) is the main lever to amortize that overhead—see the [Triton Inference Server guide](https://docs.ultralytics.com/guides/triton-inference-server/) and the [inference latency overview](https://www.ultralytics.com/glossary/inference-latency).
